[dloh2236] iP

dloh2236 opened the pull request:
# Oreo
> "Why aren't you giving me belly rubs?" - Oreo, All Day Everyday 

Oreo is your trusty canine companion and personal assistant. She's always by your side,  ready to assist. Oreo excels at keeping track of tasks and schedules, effortlessly organizing your day with her unwavering dedication. Her boundless energy and unwavering loyalty make her the perfect partner in the chaos of your daily routine. Whether it's reminding me of important deadlines or providing a comforting presence during late-night work sessions, Oreo is an indispensable part of your life.

Oreo is:
- text-based
- easy to learn
- **_SUPER FAST_**

All you need to do is:
1. download it [here](https://github.com/dloh2236/ip/releases/tag/Level-9).
2. run the jar file in your terminal using: `java -jar oreo.jar`
3. add your tasks
4. and let her do the rest 🐶 

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a Java Programmer, you can practice Java too. Here's the `main` method:
```
public static void main(String[] args) {
            Oreo oreo = new Oreo("./oreo.txt");
            oreo.run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

dloh2236 committed:
Changed chatbot name, greet and exit message

dloh2236 committed:
Changed the greet message to include a doggy

dloh2236 committed:
Oreo now echoes back command, exit message when 'bye' command input

dloh2236 committed:
Added ability to add text and display back when requested

dloh2236 committed:
Added a textformat class to clean up reply formatting by auto indenting and adding of horizontal lines

dloh2236 committed:
Created Task Class that handles Task methods and attributes
Re-did Duke file to declaring attributes and methods and main function just runs an instance of Duke

dloh2236 committed:
Checks if Task being marked is marked or unmarked already outputs a different message
If all task completes, send all complete message
Checks if mark/unmark is valid
Allow for adding of tasks called mark

dloh2236 committed:
Added Deadline, Event, ToDo classes that are child classes of Task
Implemented a way to set deadline event and to when given the appropriate command
Also able to be manipulated like a task like before

dloh2236 committed:
Implemented delete function, changed taskList to ArrayList

dloh2236 committed:
Added Exception catching of illegal commands

dloh2236 committed:
Enhanced the erroneous command to handle more errors. Should cover all wrong input of commands including the lack of commands

dloh2236 committed:
minor change to empty list message

dloh2236 committed:
Fixed several logic errors for error input, improved readability of some messages and changed handling of certain errors to more relevant methods

dloh2236 committed:
wrote text-ui-test, program passes test

dloh2236 committed:
Implement a save function that writes a .txt file onto the hardisk with items on list
reads a duke.txt if exist, if not creates one.
when bye command is inputted, duke write the task onto duke.txt file

dloh2236 committed:
Custom greet message based on if saved task exist

dloh2236 committed:
Comment run method

dloh2236 committed:
debugging merge issue for git

dloh2236 committed:
add Duke.java to commit

dloh2236 committed:
Resolve merge conflict

dloh2236 committed:
Merge conflict resolve in master

dloh2236 committed:
Resolve conflict again

dloh2236 committed:
Resolve conflict for pull

dloh2236 committed:
Fix filepath to work for with TextUiTesting

dloh2236 committed:
Implement basic date time parser when adding new tasks

dloh2236 committed:
Fixed file writing and file reader to read new time format

dloh2236 committed:
Change date format input

dloh2236 committed:
Catch all invalid time inputs

dloh2236 committed:
Merge branch 'branch-Level-8' to master

dloh2236 committed:
Clean up import list

dloh2236 committed:
Minor bug fix for list() and Text UI output

dloh2236 committed:
Merge branch 'master' into branch-A-MoreOOP to update code for branch-A-OOP

dloh2236 committed:
Implemented UI class that handles text UI of the chatbot

dloh2236 committed:
Implement Storage and TaskList Class

dloh2236 committed:
Bug fixes for some ui messages not printing
Further OOP for TaskList to handle mark/unmark delete methods

dloh2236 committed:
Implement Parser Class, Command Class and its child classes, main logic refactor

dloh2236 committed:
Merge branch 'branch-A-MoreOOP'

dloh2236 committed:
Change bot name to Oreo

dloh2236 committed:
Package and import packages accordingly

dloh2236 committed:
Merge branch 'branch-A-Packages'

dloh2236 committed:
Merge commit 'ed6d4d25bd4e34461919ffe785a6f10f4c087552' add gradle support

dloh2236 committed:
Add unit test for AddCommand Class to test for valid and invalid inputs

dloh2236 committed:
JUnit test on Task generateTask method for valid and invalid inputs

dloh2236 committed:
Merge branch 'branch-A-Junit'

dloh2236 committed:
Change acrchiveBaseName to 'oreo'

dloh2236 committed:
Implement find command for oreo

dloh2236 committed:
change boolean variable name to comply with coding standard

dloh2236 committed:
Finish coding standard correction

dloh2236 committed:
Add Javadocs to important classes

dloh2236 committed:
Merge branch 'branch-A-JavaDoc'

dloh2236 committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/oreo/task/Task.java

dloh2236 committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/oreo/task/Task.java
#	src/main/java/oreo/task/TaskList.java

dloh2236 committed:
Add javadocs to find methods

dloh2236 committed:
Minor change to message

dloh2236 committed:
Implement checkstyle build with gradle

dloh2236 committed:
Merge branch 'branch-A-CheckStyle'

dloh2236 committed:
Implment GUI and integrated logic with GUI (basic)

Chrainx commented on src/main/java/oreo/command/AddCommand.java:
```diff
@@ -0,0 +1,35 @@
+            ui.say(e.getMessage());
+        }
+    }
+}
```
Shouldn't it have java doc comment?

Chrainx commented on src/main/java/oreo/command/ByeCommand.java:
```diff
@@ -0,0 +1,11 @@
+    public void execute(Ui ui, TaskList tasks) {
+        // do nothing
+    }
+}
```
Shouldn't it have java doc comment?

Chrainx commented on src/main/java/oreo/command/Command.java:
```diff
@@ -0,0 +1,29 @@
+ * @version 03/09/2023
+ */
+public abstract class Command {
+    /**
+     * Executes the command
+     *
+     * @param ui Ui of the chatbot
+     * @param tasks Tasklist of the chatbot
+     */
+    public abstract void execute(Ui ui, TaskList tasks);
+
+    /**
+     * Checks if command is bye
+     *
+     * @return true if command is ByeCommand
+     */
```
Shouldn't the comments have full stop at the end

Chrainx commented on src/main/java/oreo/command/DeleteCommand.java:
```diff
@@ -0,0 +1,23 @@
+import oreo.ui.Ui;
+import java.util.Scanner;
+
+public class DeleteCommand extends Command {
+    private Scanner tokeniser;
+
+    public DeleteCommand(Scanner tokeniser) {
+        this.tokeniser = tokeniser;
+    }
+
+    @Override
+    public void execute(Ui ui, TaskList tasks) {
+        try {
+            ui.say(tasks.deleteTask(tokeniser));
+        } catch (IllegalCommandException e) {
+            ui.say(e.getMessage());
+        }
+    }
+}
```
Shouldn't it have java doc comment?

Chrainx commented on src/main/java/oreo/command/EmptyCommand.java:
```diff
@@ -0,0 +1,11 @@
+import oreo.task.TaskList;
+import oreo.ui.Ui;
+
+public class EmptyCommand extends Command {
+    @Override
+    public void execute(Ui ui, TaskList tasks) {
+        ui.say("uhhh what???");
+    }
+}
```
Shouldn't it have java doc comment?

Chrainx commented on src/main/java/oreo/command/FindCommand.java:
```diff
@@ -0,0 +1,24 @@
+
+import java.util.Scanner;
+
+public class FindCommand extends Command {
+    private Scanner tokeniser;
+
+    public FindCommand(Scanner tokeniser) {
+        this.tokeniser = tokeniser;
+    }
+
+    @Override
+    public void execute(Ui ui, TaskList tasks) {
+        try {
+            ui.say(tasks.findTasksWith(tokeniser));
+        } catch (IllegalCommandException e) {
+            ui.say(e.getMessage());
+        }
+    }
+}
```
Shouldn't it have java doc comment?

Chrainx commented on src/main/java/oreo/command/ListCommand.java:
```diff
@@ -0,0 +1,12 @@
+import oreo.task.TaskList;
+import oreo.ui.Ui;
+
+public class ListCommand extends Command {
+
+    @Override
+    public void execute(Ui ui, TaskList tasks) {
+        ui.say(tasks.list());
+    }
+}
```
Shouldn't it have java doc comment? I have seen this issue in many other places

Chrainx commented on src/main/java/oreo/task/Deadline.java:
```diff
@@ -0,0 +1,40 @@
+
+    private String time;
+
+    public Deadline(String d, String date, String time) {
```
Shouldn't the d be named into a more straightforward variable?

ravern replied:
Adding on here — I think that while I can inspect the superclass `Task` to figure out what `d` means, it's an extra step that might slow down others working on your code.

Chrainx commented on src/main/java/oreo/task/Event.java:
```diff
@@ -0,0 +1,57 @@
+    private String fromTime;
+    private String toDate;
+    private String toTime;
+    public Event(String d, String fromDate, String fromTime,
```
Shouldn't the d be named into a more straight forward variable? I have seen this issue in many other places too.

Chrainx reviewed (commented):
In overall, You did a very good job in following the coding standard and only have little misses.

ravern commented on src/main/java/oreo/Oreo.java:
```diff
@@ -0,0 +1,91 @@
+        /*
+        This portion displays exit message
+         */
+        ui.sayBye();
```
This isn't a direct violation on anything taught within the course so far, but IMO the `/* hello */` style of comments do affect the readability of your code for reviewing. Perhaps the `// hello` style of comments are more suited for inline commenting of code.

ravern commented on src/main/java/oreo/command/AddCommand.java:
```diff
@@ -0,0 +1,35 @@
+            Task newTask = Task.generateTask(command, tokeniser);
+            tasks.add(newTask);
+            ui.say("Gotchu! noted down: \n"
+                    + Ui.indentLineBy(newTask.toString(), 2)
```
Cool method!

ravern commented on src/main/java/oreo/exception/IllegalCommandException.java:
```diff
@@ -0,0 +1,18 @@
+
+    public IllegalCommandException(String cmsg) {
+        this.cmsg = cmsg;
+    }
```
I think the context here is a little wide to use a shortened form `cmsg`. I can't tell what `cmsg` stands for here even after reading a little further. Command message?

ravern commented on src/main/java/oreo/exception/IllegalDateTimeException.java:
```diff
@@ -0,0 +1,16 @@
+
+import java.time.DateTimeException;
+
+public class IllegalDateTimeException extends Exception {
+    private String msg;
+
+    public IllegalDateTimeException(String msg) {
+        this.msg = msg;
+    }
+
+    @Override
+    public String getMessage() {
+        return msg;
+    }
+}
```
For `Exception` subclasses, I believe you can simply use `message` which is already available as part of the `Exception` superclass?

```java
public class IllegalDateTimeException extends Exception {
    public IllegalDateTimeException(String msg) {
        super(msg);
    }
}
```

With this, `getMessage` _just works._

ravern commented on src/main/java/oreo/parser/Parser.java:
```diff
@@ -0,0 +1,46 @@
+        }
+
+        switch (command) {
+        case("bye") :
+            return new ByeCommand();
+        case("list") :
+            return new ListCommand();
+        case("unmark") :
+        case("mark") :
+            return new MarkUnmarkCommand(command, tokeniser);
+        case("delete") :
+            return new DeleteCommand(tokeniser);
+        case("find") :
+            return new FindCommand(tokeniser);
+        default:
+            return new AddCommand(command, tokeniser);
```
Are the parentheses around the strings on the case labels needed here?

ravern commented on src/main/java/oreo/task/Task.java:
```diff
@@ -0,0 +1,165 @@
+        return this.isComplete;
+    }
+
+    /**
+     * Change mark to opposite of the current status.
+     */
+    public void switchMark() {
+        isComplete = !isComplete;
+    }
```
Perhaps a better name here is `toggleMark`. `switchMark` made me think of switching _to_ something, but there's no parameter to specify what to switch to which confused me at first.

ravern reviewed (commented):
Great work! Abstractions are quite sound and kept simple, making this PR a breeze to review. Thanks!

dloh2236 committed:
Improve GUI for Oreo

dloh2236 committed:
Merge branch 'branch-Level-10'

dloh2236 committed:
Build gradle file: change mainClass set to new .java file after GUI integration

Change as now Launcher.java is the main file that starts the application

Simple change in file name for main file

dloh2236 committed:
Parser Class: pass invalid command to a AddCommand instance where exceptions are further manage

Parser Class and Task Class: This cause clutter in the Task File as it has to handle invalid commands which is not relevant to that Class

Create a new Command Subclass called InvalidCommand that throws the exception when execute() is called. Parser Class now returns a InvalidCommand Exception as its default case

dloh2236 committed:
Add assertions to document important assumptions that should hold

dloh2236 committed:
Merge pull request #1 from dloh2236/branch-A-Assertions

Add assertions to document important assumptions that should hold

dloh2236 committed:
Create CI workflow

dloh2236 committed:
Merge branch 'branch-A-CI'

dloh2236 committed:
Remove I/O redirection test

dloh2236 committed:
Correct .yml file

dloh2236 committed:
Refactor Task Class and TaskList Class

Task Class: Improve code quality by reducing nested if-else statements to more methods to handle sub cases

Tasklist Class: Break if-else statement up to more logical flow, catch erroroneous input first before handling proper data

dloh2236 committed:
Merge pull request #3 from dloh2236/branch-A-CodeQuality

Refactor Task Class and TaskList Class

dloh2236 committed:
Add edit/update/modify function

Oreo Class: Add cache to pass information between scenes
Command Class: Implements a new method that execute task while in edit mode
Parser Class: Parse new command, parse command in edit mode
MainWindow Class: Implement new method that handles edits (new edit button/textfield calls handleEditInput()). Implement handler that transits chat from normal user input mode to edit mode, cahe command and task to edit
Task Classes: Implement methods prints in edit format for user to edit.

dloh2236 committed:
UI improvements: Dialogbox now has different colour for oreo
Change how different task is displayed
MainWindow now has a edit mode with different textfield and button to call a different function

dloh2236 committed:
Merge pull request #4 from dloh2236/branch-C-Update

Add edit/update/modify function

dloh2236 committed:
Merge pull request #5 from dloh2236/branch-A-BetterGui

Branch a better gui

dloh2236 committed:
Merge branch 'master' of https://github.com/dloh2236/ip

dloh2236 committed:
Implement mark/unmark all and delete all/clear

dloh2236 committed:
Merge branch 'branch-Self-AllOps'

dloh2236 committed:
Minor Change for ui class to formatter class as it no longer handles ui. Refactoring of rest of the classes who use methods in the Formatter class

dloh2236 committed:
Refactor TaskList class to let command processing be done by command class

dloh2236 committed:
Implement help command, general help and more specific help

dloh2236 committed:
Merge branch 'branch-C-Help'

dloh2236 committed:
Improve time parser to accept more date and time formats and different combinations. Now date time formatting is more natural and easy to use for users

dloh2236 committed:
Merge branch 'branch-C-NaturalDates'

dloh2236 committed:
Minor change, refactor to message output and code

dloh2236 committed:
UI improvments

dloh2236 committed:
Allow more inputs of datetime such as "22 sep" no need for "22 Sep"
Refactor date time parsing mechanism so error messages are more specific
Refactor get next date of time occurence to consider event ending times

dloh2236 committed:
Further improvments on command format error message and checking process

dloh2236 committed:
Edit help txt files for grammar, spelling and made some easier and more useful to users

dloh2236 committed:
Add javadocs

dloh2236 committed:
Added more unit test

dloh2236 committed:
Product screenshot

dloh2236 committed:
A-UserGuide

Added User Guide

dloh2236 committed:
Merge pull request #6 from dloh2236/branch-A-UserGuide

A-UserGuide

dloh2236 committed:
Minor changes

dloh2236 committed:
Minor bug fix to file reading and UI improvement to fix width and allow flexible height

dloh2236 committed:
New User Guide

dloh2236 committed:
Update product image

dloh2236 committed:
Add more date formats to parse, some bug fixes with regards to date time format

dloh2236 committed:
Update and rectified README and help.

dloh2236 committed:
Update README.md

dloh2236 committed:
Update README.md

dloh2236 committed:
Update README.md

dloh2236 committed:
Update README.md

dloh2236 committed:
Minor Bug fixes
- Help command handles wrong commands like a champ now
- DateTime parsing logic for events is now improved - oreo can math out the dates better
- Fixed DateTime parser for all Java 11!

dloh2236 committed:
Update README.md to reflect datetime parser fixed for all java 11

dloh2236 committed:
Update README.md for latest version 1.0.1

dloh2236 committed:
Update README.md test for GIthub Pages

dloh2236 committed:
Update README.md format for github pages

dloh2236 committed:
Update README.md test for Github Pages

dloh2236 committed:
Update README.md again for github pages formatting issues

dloh2236 committed:
Update README.md

dloh2236 committed:
Update README.md

dloh2236 committed:
Merge branch 'master' of https://github.com/dloh2236/ip

dloh2236 committed:
Update README.md

dloh2236 committed:
Update README.md

dloh2236 committed:
Update README format for Github Pages
